lib/libvarnishapi/vut.c

```diff
@@ -382,17 +383,44 @@ VUT_Fini(struct VUT **vutp)
 	FREE_OBJ(vut);
 }
 
+static void
+vut_CursorError(struct VUT *vut, vtim_mono *last)
+{
+	const char *diag;
+	vtim_mono now;
+
+	CHECK_OBJ_NOTNULL(vut, VUT_MAGIC);
+	AN(vut->vsl);
+	AN(last);
+
+	diag = VSL_Error(vut->vsl);
+	if (diag == NULL)
+		diag = "Missing diagnostic";
+
+	now = VTIM_mono();
+	if (isnan(*last) || *last + 1 < now) {
+		fprintf(stderr, "Failed to acquire log: %s\n", diag);
+		*last = now;
+	}
+}
+
 int
 VUT_Main(struct VUT *vut)
 {
 	struct VSL_cursor *c;
 	int i = -1;
-	int hascursor = -1;
+	int hascursor = -1, trycursor = 0;
+	vtim_mono failcursor = NAN;
 
 	CHECK_OBJ_NOTNULL(vut, VUT_MAGIC);
 	AN(vut->vslq);
 
 	while (!VSIG_int && !VSIG_term) {
+		if (trycursor == 10) {
```

Comment:
We already have the `-t` option that does something similar for the VSM, the `trycursor` limit is only for the VSL once we are already "connected" to a varnish server. Or maybe I'm missing something else?

Reply:
I have not tried this myself, but could this be relevant for crashes, when the VSM stays intact?

Reply:
I am uncertain of this feature. With this patch, we'd allow a maximum of 1 second (10 * 0.1s) before the log section reappears. I worry that for our very large persistent caches, then that may not be enough. Also I believe the OS may take more than a sec to clean up the old child process when handling a very large memory footprint process like the Varnish child, especially so if there was swap usage involved. Perhaps some logging to stderr would be better, but otherwise continue to retry indefinetely?

Reply:
My concern today is that a the purely textual diagnostic doesn't tell you whether the error is recoverable or not. For example, X attempts at acquiring a corrupted VSL should be enough to leave because you'll likely loop forever. If the child is down, then you would eventually recover. We probably need a way to know the diagnostic severity to exit the loop.

Reply:
To add facts to this discussion:
```
$ varnishadm param.set auto_restart false
$ varnishlog &
[2] 13430
$ kill 13188
Log abandoned (vsl)
Log abandoned (vsm)
Failed to acquire log: No VSL chunk found (child not started ?)
Log abandoned (vsm)
Log abandoned (vsm)
Log abandoned (vsm)
Log abandoned (vsm)
Log abandoned (vsm)
Log abandoned (vsm)
Log abandoned (vsm)
Log abandoned (vsm)
Log abandoned (vsm)
Failed to acquire log too many times

[2]+  Exit 1                  varnishlog
```
The varnishlog exit happened "immediately", after 1 second.

So my verdict has changed to a clear _no_. We should not have this short, fixed timeout. We might want to rate limit diagnostic output (I like exponential backoff, waving to @slimhazard ), and we might want to offer a parameter for the maximum time to wait, but this definitely is not a good default (and in particular not fixed) behavior.

Reply:
also note the documentation of `-t`:
> -t <seconds|off> ... Defaults to 5 seconds.

Reply:
bugwash: remove from PR, come back to it later ifneedbe